src/passes/pass.cpp

```diff
@@ -811,4 +810,25 @@ int PassRunner::getPassDebug() {
   return passDebug;
 }
 
+bool PassRunner::passRemovesDebugInfo(const std::string& name) {
+  return name == "strip" || name == "strip-debug" || name == "strip-dwarf";
+}
+
+bool PassRunner::shouldPreserveDWARF() {
+  // Check if the debugging subsystem wants to preserve DWARF.
+  if (!Debug::shouldPreserveDWARF(options, *wasm)) {
+    return false;
+  }
+
+  // We may need DWARF. Check if one of our previous passes would remove it
+  // anyhow, in which case, there is nothing to preserve.
+  for (auto& pass : passes) {
+    if (passRemovesDebugInfo(pass->name)) {
+      return false;
+    }
+  }
```

Comment:
Is this called a lot?  Is it worth caching the result?

Reply:
I don't think it's called much, as it's limited by the number of passes. But I suppose it's worth caching, even though it adds some complexity. I added that now, and also some assertions to avoid possible issues.

Reply:
No worries.  Looks good as is then.